pkg/pillar/types/downloadertypes.go

```diff
@@ -167,3 +168,71 @@ func AllowNonFreePort(gc ConfigItemValueMap, objType string) bool {
 		return false
 	}
 }
+
+// AppImgResolveConfig key/index to this is the ImageID which is allocated by the controller.
+// It will resolve container image tag to sha256
+type AppImgResolveConfig struct {
+	AppID            uuid.UUID
+	ImageID          uuid.UUID
+	DatastoreID      uuid.UUID
+	Name             string
+	NameIsURL        bool // If not we form URL based on datastore info
+	AllowNonFreePort bool
+	PurgeCounter     uint32
+}
+
+// Key : ImageID is used to differentiate different config
+func (config AppImgResolveConfig) Key() string {
+	return fmt.Sprintf("%s-%s-%v", config.AppID.String(), config.ImageID.String(), config.PurgeCounter)
+}
```

Comment:
The client is asking to resolve a name for some particular datastore, such as nginx:latest in a docker datastore. That is fundamentally the key. The fact that it might be used for app instance and some imageid has nothing to do with the resolution AFAICT.

Reply:
Resolved.